text/0113-secure-script-container.md

```diff
@@ -0,0 +1,183 @@
+component encapsulation via shadow DOM it was possible for library authors to:
+
+- Listen for events and know who fired them
+- Define global styling rules that would get applied on every element
+- Query the entire document for any element on the page
+
+These libraries are typically added as global scripts to the document root rather than deeper 
+integrations into the components themselves. When every component's root element is a shadow
+root, there is no light DOM for them to work with, and their functionality is severely crippled.
+
+Some examples of these libraries are:
+
+- **Analytics tools**: Google Tag Manager / Google Analytics, Pendo
+- **Personalization platforms**: Optimizely, Evergage, Google Optimize
+- **Instrumentation tools**: New Relic, Sentry.io
+
+In addition, developers are finding that this model is not compatible with their **design systems** and **testing frameworks**.
+
+The goal of this proposal is to provide a way for application developers to continue to use their existing integrations until library authors can support web component based applications.
+
+## Design
+
+`secure-script` will solve for the synthetic shadow and subsequently the native shadow.
+
+### Synthetic Shadow
+Many LWC applications apply the [synthetic-shadow polyfill](https://github.com/salesforce/lwc/tree/master/packages/%40lwc/synthetic-shadow/). This polyfill emulates the native shadow behavior while still allowing global styles to cascade into the shadow trees. Synthetic shadow currently [patches](https://github.com/salesforce/lwc/tree/master/packages/%40lwc/synthetic-shadow/src/polyfills/document-shadow) the `document.querySelector` and `document.querySelectorAll` APIs which is what prevents developers from penetrating the shadow DOM boundaries either in their LWC components or any other scripts added to the document. We only want to solve for the latter scenario.
+
+To ensure we can solve the problem while continuing to guarantee the benefits that are provided by the synthetic-shadow polyfill for all other components, we are currently proposing the following characteristics for the secure-script design:
+
+* Create a secure JS sandbox which doesn’t allow its object graph to be extended using [`Object.preventExtensions`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/preventExtensions)
+  * The sandbox is created using a same-domain iframe provided by Locker v.Next
+  * This has to be done before the synthetic shadow is imported
+
+* Run or include the third party library script inside this sandbox
+  * This allows the code to run in a non-patched environment, while the rest of the document is running synthetic shadow
+* Package the above functionality in a web component which semantically matches the `<script>` tag.
+
+
+### Native Shadow
+
+For applications which are using the native shadow, the underlying implementation will be different. There are two approaches we can choose from:
```

Comment:
We should add way more detail here about how we are thinking on implementing this as we move forward

Reply:
I agree, I think it's important to understand that now, even if its implementation comes later.

Reply:
will do in a bit!

Reply:
added... please review